### DISCUSSION TITLE
Can signing in with `redirect: false` be combined with secured pages?

### DISCUSSION THREAD

**Discussion by grazianodev** [nextauthjs/next-auth#9784] [Help]
### Question 💬

I have a Nextjs 13 frontend (app router) where I'm trying to redirect a user, who successfully logged in with email+password, to a dashboard page, which is secured with the `withAuth` middleware. I do not want a page reload but a client-side route change, so I am using the `redirect: false` option for the sign in function, and I have also set a callback url:

```
const response = await signIn( 'credentials', {
   email: data.email,
   password: data.password,
   callbackUrl: '/dashboard',
   redirect: false 
})

if( response.error ) {
   // handle error
}

router.push( response.url )
```

I have an issue with the following scenario: 

1) A non-logged in user visits the dashboard page; 
2) The dashboard page, being secured via middleware, redirects him to the login page (with the dashboard url appended); 
3) The user logs in, cookies are successfully set
4) The login page redirects to itself instead of redirecting to `/dashboard`

Everything "works" as expected if I don't use `redirect: false`, or if I do use it but remove the middleware and unsecure the dashboard; by the way, I have the same issue if I protect the dashboard page with `getServerSession` rather than the middleware. I have read other similar issues but the solution there was to add a callback url, which I have already done. 

### How to reproduce ☕️

Here's a [minimal reproduction](https://github.com/grazianodev/next-auth-test), I would appreciate any help! 

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by jimmy-qian** [comment]
I have a similar issue!
Almost the exact setup, using `redirect: false` to handle error on the same page, but then `callbackUrl` is no longer used for redirecting when `signIn()` returns `200`.

**Comment by stale** [comment]
It looks like this issue did not receive any activity for 60 days. It will be closed in 7 days if no further activity occurs. If you think your issue is still relevant, commenting will keep it open. Thanks!

**Comment by shazsyed** [comment]
I ran into similar issue, where I am showing error in toast notification and don't want a redirect `(redirect: false)` but now the `callbackUrl` is not working. I am using next router as a workaround, it would be great to have `callbackUrl` working with `redirect: false`

**Comment by pepew-le-boss** [comment]
Any solution on this subject? Same issue here

**Comment by ShahriarKh** [comment]
I also agree that this should work out-of-the-box, but until it's fixed, here's a workaround:
```jsx
const router = useRouter(); // import from next/navigation

function handleLogin(data) {
  signIn("credentials", {
    username: data.username,
    password: data.password,
    redirect: false, // don't redirect to error page
  }).then((res) => {
    if (res.ok) {
      // your success message
      router.push(
        `${new URLSearchParams(window.location.search).get("callbackUrl")}`
      );
    } else if (res.error) {
      // your error message
    }
  });
}

```
